sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java

```diff
@@ -126,34 +125,93 @@ public boolean formatArgs(Function function) {
             this.name = name;
-            this.targetType = targetType;
         }
 
-        void checkPreconditions(List<DataType> dataTypes) {
-            DataType convertFrom = dataTypes.get(0);
-            if (!convertFrom.isConvertableTo(targetType)) {
-                throw new ConversionException(convertFrom, targetType);
+        @Override
+        public FunctionImplementation getForTypes(List<DataType> dataTypes) {
+            DataType<?> targetType;
+            if (dataTypes.size() > 1) {
+                var targetTypeViaArg = dataTypes.get(1);
+                targetType = targetTypeViaArg.id() != UndefinedType.ID ? targetTypeViaArg : type;
+            } else {
+                targetType = type;
+            }
+
+            DataType<?> sourceType = dataTypes.get(0);
+            if (!sourceType.isConvertableTo(targetType)) {
+                throw new ConversionException(sourceType, targetType);
             }
+
+            return new CastFunction(
+                new FunctionInfo(new FunctionIdent(name, dataTypes), targetType),
+                castOnNormalizeException,
+                castOnEvaluateException
+            );
+        }
+    }
+
+    private static class TryCastResolver extends BaseFunctionResolver {
+
+        private final DataType<?> type;
+        private final String name;
+
+        private final BiFunction<Symbol, DataType<?>, Symbol> castOnNormalizeException =
+            (argument, returnType) -> Literal.NULL;
+        private final BiFunction<Object, DataType<?>, Object> castOnEvaluateException =
+            (argument, returnType) -> null;
+
+        TryCastResolver(DataType type, String name) {
+            super(FuncParams.builder().withIndependentVarArgs(Param.ANY).build());
+            this.type = type;
+            this.name = name;
         }
 
         @Override
         public FunctionImplementation getForTypes(List<DataType> dataTypes) {
-            checkPreconditions(dataTypes);
-            return new CastFunction(new FunctionInfo(new FunctionIdent(name, dataTypes), targetType));
+            DataType<?> targetType;
+            if (dataTypes.size() > 1) {
+                var targetTypeViaArg = dataTypes.get(1);
+                targetType = targetTypeViaArg.id() != UndefinedType.ID ? targetTypeViaArg : type;
+            } else {
+                targetType = type;
+            }
+
+            return new CastFunction(
+                new FunctionInfo(new FunctionIdent(name, dataTypes), targetType),
+                castOnNormalizeException,
+                castOnEvaluateException
+            );
         }
     }
 
     public static void register(ScalarFunctionModule module) {
-        for (Map.Entry<String, DataType> function : CastFunctionResolver.CAST_SIGNATURES.entrySet()) {
-            module.register(function.getKey(), new Resolver(function.getValue(), function.getKey()));
+        for (Map.Entry<String, DataType> function : CAST_SIGNATURES.entrySet()) {
```

Comment:
Do we still need the `CAST_SIGNATURES` in the current form? I thought we're going to remove the `to_<typeName>` notation. Or is it still required for BWC?

Reply:
It is actually needed for bwc.

Reply:
Could you add a comment about this? Also noting at which point we can remove it.